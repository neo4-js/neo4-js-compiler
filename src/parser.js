// @flow

import nearley from "nearley";
import grammar from "./grammar";

const g = nearley.Grammar.fromCompiled(grammar);

export function parse(str: string) {
  const parser = new nearley.Parser(g);
  parser.feed(str);
  return parser.results[0];
}
