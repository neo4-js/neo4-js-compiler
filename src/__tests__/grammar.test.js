import { parse } from "../parser.js";

describe("parser", () => {
  it("should parse simplest possible model", () => {
    let schema = "type User { name: String }";

    try {
      const result = parse(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });

  it("should parse model without properties", () => {
    let schema = "type User { }";

    try {
      const result = parse(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });

  it("should parse schema with two models", () => {
    let schema = `
      type User {
        name: String
      };

      type Task {
        title: String;
        priority: Number
        done: Boolean,
      };
    `;

    try {
      const result = parse(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });

  it("should parse implements", () => {
    let schema = `
      type Node {
        id: String!
      }

      type User implements Node {
        name: String!
      }

      type Task implements Node {
        title: String!
      }
    `;

    try {
      const result = parse(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });

  it("should parse schema with simple relation", () => {
    let schema = `
      type User {
        name: String!
        tasks: has many Task using UserTaskRelation
      }

      type Task {
        title: String!,
        creator: has one User using UserTaskRelation
      }

      relation UserTaskRelation from User to Task via "created" {
        order: Number
      }
    `;

    try {
      const result = parse(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });

  it("should parse complex model", () => {
    let schema = `
      type Node {
        id: String!
      }

      type User implements Node {
        name: String!
        tasks: has many Task using UserCreatedTasks
        assignedTasks: has many Task using UserAssignedTasks
        age: Number
      }

      type Task {
        name: String!
        creator: has one User using UserCreatedTasks
        assignee: has one User using UserAssignedTasks,
        done: Boolean,
      }

      relation UserCreatedTasks from User to Task via "created" { }

      relation UserAssignedTasks from User to Task via "assigned" {
        assignedDate: String!
      }
    `;

    try {
      const result = parse(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });

  it("should parse default values", () => {
    let schema = `
      type User {
        name: String = "test",
        meh: Number = 3e4
      }

      type Task {
        name: String!;
        done: Boolean = false,
        order: Number = 394
      }
    `;

    try {
      const result = parse(schema);
      expect(result).toMatchSnapshot();
    } catch (err) {
      expect(err).toBeFalsy();
    }
  });
});
