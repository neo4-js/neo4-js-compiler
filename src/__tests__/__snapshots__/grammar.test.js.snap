// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser should parse complex model 1`] = `
Array [
  Object {
    "name": "Node",
    "properties": Array [
      Object {
        "name": "id",
        "required": true,
        "scalar": true,
        "type": "String",
      },
    ],
    "type": "Model",
  },
  Object {
    "implements": "Node",
    "name": "User",
    "properties": Array [
      Object {
        "name": "name",
        "required": true,
        "scalar": true,
        "type": "String",
      },
      Object {
        "direction": null,
        "many": true,
        "name": "tasks",
        "relation": "UserCreatedTasks",
        "to": "Task",
        "type": "Relation",
      },
      Object {
        "direction": null,
        "many": true,
        "name": "assignedTasks",
        "relation": "UserAssignedTasks",
        "to": "Task",
        "type": "Relation",
      },
      Object {
        "defaultValue": null,
        "name": "age",
        "scalar": true,
        "type": "Number",
      },
    ],
    "type": "Model",
  },
  Object {
    "name": "Task",
    "properties": Array [
      Object {
        "name": "name",
        "required": true,
        "scalar": true,
        "type": "String",
      },
      Object {
        "direction": null,
        "many": false,
        "name": "creator",
        "relation": "UserCreatedTasks",
        "to": "User",
        "type": "Relation",
      },
      Object {
        "direction": null,
        "many": false,
        "name": "assignee",
        "relation": "UserAssignedTasks",
        "to": "User",
        "type": "Relation",
      },
      Object {
        "defaultValue": null,
        "name": "done",
        "scalar": true,
        "type": "Boolean",
      },
    ],
    "type": "Model",
  },
  Object {
    "from": "User",
    "name": "UserCreatedTasks",
    "properties": Array [],
    "to": "Task",
    "type": "Relation",
    "via": "created",
  },
  Object {
    "from": "User",
    "name": "UserAssignedTasks",
    "properties": Array [
      Object {
        "name": "assignedDate",
        "required": true,
        "scalar": true,
        "type": "String",
      },
    ],
    "to": "Task",
    "type": "Relation",
    "via": "assigned",
  },
]
`;

exports[`parser should parse default values 1`] = `
Array [
  Object {
    "name": "User",
    "properties": Array [
      Object {
        "defaultValue": Object {
          "type": "String",
          "value": "test",
        },
        "name": "name",
        "scalar": true,
        "type": "String",
      },
      Object {
        "defaultValue": Object {
          "type": "Number",
          "value": 30000,
        },
        "name": "meh",
        "scalar": true,
        "type": "Number",
      },
    ],
    "type": "Model",
  },
  Object {
    "name": "Task",
    "properties": Array [
      Object {
        "name": "name",
        "required": true,
        "scalar": true,
        "type": "String",
      },
      Object {
        "defaultValue": Object {
          "type": "Boolean",
          "value": "false",
        },
        "name": "done",
        "scalar": true,
        "type": "Boolean",
      },
      Object {
        "defaultValue": Object {
          "type": "Number",
          "value": 394,
        },
        "name": "order",
        "scalar": true,
        "type": "Number",
      },
    ],
    "type": "Model",
  },
]
`;

exports[`parser should parse implements 1`] = `
Array [
  Object {
    "name": "Node",
    "properties": Array [
      Object {
        "name": "id",
        "required": true,
        "scalar": true,
        "type": "String",
      },
    ],
    "type": "Model",
  },
  Object {
    "implements": "Node",
    "name": "User",
    "properties": Array [
      Object {
        "name": "name",
        "required": true,
        "scalar": true,
        "type": "String",
      },
    ],
    "type": "Model",
  },
  Object {
    "implements": "Node",
    "name": "Task",
    "properties": Array [
      Object {
        "name": "title",
        "required": true,
        "scalar": true,
        "type": "String",
      },
    ],
    "type": "Model",
  },
]
`;

exports[`parser should parse model without properties 1`] = `
Array [
  Object {
    "name": "User",
    "properties": Array [],
    "type": "Model",
  },
]
`;

exports[`parser should parse schema with simple relation 1`] = `
Array [
  Object {
    "name": "User",
    "properties": Array [
      Object {
        "name": "name",
        "required": true,
        "scalar": true,
        "type": "String",
      },
      Object {
        "direction": null,
        "many": true,
        "name": "tasks",
        "relation": "UserTaskRelation",
        "to": "Task",
        "type": "Relation",
      },
    ],
    "type": "Model",
  },
  Object {
    "name": "Task",
    "properties": Array [
      Object {
        "name": "title",
        "required": true,
        "scalar": true,
        "type": "String",
      },
      Object {
        "direction": null,
        "many": false,
        "name": "creator",
        "relation": "UserTaskRelation",
        "to": "User",
        "type": "Relation",
      },
    ],
    "type": "Model",
  },
  Object {
    "from": "User",
    "name": "UserTaskRelation",
    "properties": Array [
      Object {
        "defaultValue": null,
        "name": "order",
        "scalar": true,
        "type": "Number",
      },
    ],
    "to": "Task",
    "type": "Relation",
    "via": "created",
  },
]
`;

exports[`parser should parse schema with two models 1`] = `
Array [
  Object {
    "name": "User",
    "properties": Array [
      Object {
        "defaultValue": null,
        "name": "name",
        "scalar": true,
        "type": "String",
      },
    ],
    "type": "Model",
  },
  Object {
    "name": "Task",
    "properties": Array [
      Object {
        "defaultValue": null,
        "name": "title",
        "scalar": true,
        "type": "String",
      },
      Object {
        "defaultValue": null,
        "name": "priority",
        "scalar": true,
        "type": "Number",
      },
      Object {
        "defaultValue": null,
        "name": "done",
        "scalar": true,
        "type": "Boolean",
      },
    ],
    "type": "Model",
  },
]
`;

exports[`parser should parse simplest possible model 1`] = `
Array [
  Object {
    "name": "User",
    "properties": Array [
      Object {
        "defaultValue": null,
        "name": "name",
        "scalar": true,
        "type": "String",
      },
    ],
    "type": "Model",
  },
]
`;
